src/payjoin/app.rs

```diff
@@ -0,0 +1,214 @@
+            Ok(recv_res.map_err(|e| Error::Server(e.into()))?)
+        })?;
+
+        tracing::trace!("check2");
+        // Receive Check 3: receiver can't sign for proposal inputs
+        let proposal = proposal.check_no_mixed_input_scripts()?;
+        tracing::trace!("check3");
+
+        // Receive Check 4: have we seen this input before? More of a check for non-interactive i.e. payment processor receivers.
+        let payjoin = proposal.check_no_inputs_seen_before(|input| {
+            // TODO implement input_seen_before database check
+            // Ok(!self.insert_input_seen_before(*input).map_err(|e| Error::Server(e.into()))?)
+            Ok(false)
+        })?;
+        tracing::trace!("check4");
+
+        // Receive Check 4: receiver can't sign for proposal inputs
+        let (tx, rx) = mpsc::channel::<(ScriptBuf, Sender<Result<bool, ApplicationError>>)>();
+        let tx = Arc::new(Mutex::new(tx));
+
+        let app = self.app.clone();
+        let profile = self.profile.clone();
+        let network = network.clone();
+        tokio::spawn(async move {
+            while let Ok((input, response_tx)) = rx.recv() {
+                let result = check_script_owned(&app, &profile, network, &input).await;
+                let _ = response_tx.send(result);
+            }
+        });
+
+        let mut provisional_payjoin = payjoin.identify_receiver_outputs(|output_script| {
+            let (response_tx, response_rx) = mpsc::channel::<Result<bool, ApplicationError>>();
+            {
+                let tx = tx.lock().expect("lock"); // TODO Handle lock error if needed
+                tx.send((output_script.to_owned(), response_tx))
+                    .map_err(|e| Error::Server(e.into()))?;
+            }
+            let recv_res = response_rx.recv().map_err(|e| Error::Server(e.into()))?;
+            Ok(recv_res.map_err(|e| Error::Server(e.into()))?)
+        })?;
+
+        // Don't throw an error. Continue optimistic process even if we can't contribute inputs.
+        self.try_contributing_inputs(&mut provisional_payjoin)
+            .await
+            .map_err(|e| tracing::warn!("Failed to contribute inputs: {}", e));
+
+        // Output substitution could go here
+
+        let payjoin_proposal = provisional_payjoin.finalize_proposal(
+            |psbt: &bitcoin::psbt::Psbt| {
+                Err(Error::Server(anyhow!("TODO sign psbt").into()))
+                // TODO sign proposal psbt with our inputs & subbed outputs e.g.:
+                //
+                // bitcoind
+                //     .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, Some(false))
+                //     .map(|res| bitcoin::psbt::Psbt::from_str(&res.psbt).map_err(|e| Error::Server(e.into())))
+                //     .map_err(|e| Error::Server(e.into()))?
+            },
+            None, // TODO set to bitcoin::FeeRate::MIN or similar
+        )?;
```

Comment:
How can the payjoin app submit a psbt for signing? I'm guessing this is where either `submit_payout` / `trigger_payout` would need to be called with the logic that a payjoin preempts an existing batch to include its contents in a payjoin. If it's preempting a batch, coin contributions will need to include those batch inputs as well.